# COMP-240 Homework 5

This homework assignment focuses on both data transformations and data grouping to produce aggregate results. No special libraries are required other than NumPy and Pandas.

In [15]:
import numpy as np
import pandas as pd
import requests

## Exercise 1

### Design a function that given a person's height (in m) and weight (in kg), the function returns the person's BMI score using the following equation:

$$
BMI = \frac{weight}{height^2}
$$

In [3]:
def bmi(weight, height):
    #your code goes here
    try:
        return weight / (height ** 2)
    except ZeroDivisionError:
        return 'Height cannot be 0'

bmi(57, 1.55) #example input and output 23.72

23.725286160249738

### Design another function that given the bmi returns the bmi classification for that person by adopting the following interpretation:

0 < bmi < 18.5    Underweight

18.5 <= bmi < 25  Normal

25 <= bmi < 30    Overweight

30 <= bmi         Obese 

In [32]:
def bmi_classification(bmi):
    #your code goes here
    if bmi < 0:
        return 'Bmi cannot be lower than 0.'
    elif bmi < 18.5:
        return 'Underweight'
    elif bmi < 25:
        return 'Normal'
    elif bmi < 30:
        return 'Overweight'
    else:
        return 'Obese'
    
bmi_classification(23.73) # example input

'Normal'

### Using the `map` function and lambda classify the given bmi scores by using the function you have previously designed

In [13]:
bmi_scores = [18.1, -3., 23.73, 34.32, 28.7, 22.8] # example input
#your code goes here
result = map(lambda x : bmi_classification(x), bmi_scores)
print(*result)

Underweight Bmi cannot be lower than 0. Normal Obese Overweight Normal


### From the new list, remove the values that do not have a valid interpretation

In [14]:
#your code goes here
bmi_scores_valid = filter(lambda x: x > 0, bmi_scores)
print(*bmi_scores_valid)

18.1 23.73 34.32 28.7 22.8


### Download the following dataset capturing the height and weight of 500 people into a pandas DataFrame

dataset in `csv` format can be found using the following url:

`https://storage.googleapis.com/comp240-stores/bmi.csv`

In [17]:
#your code goes here
url = 'https://storage.googleapis.com/comp240-stores/bmi.csv'
df = pd.read_csv(url)
df.head()

Gender  Height  Weight
0    Male     174      96
1    Male     189      87
2  Female     185     110
3  Female     195     104
4    Male     149      61

If you look closely, you will see that the height is given in cm. 

### Add a new attribute to the DataFrame named `HeightM` that captures for each person their height in meters

In [18]:
#your code goes here
df['HeightM'] = df['Height']/100
df.head()

Gender  Height  Weight  HeightM
0    Male     174      96     1.74
1    Male     189      87     1.89
2  Female     185     110     1.85
3  Female     195     104     1.95
4    Male     149      61     1.49

### Add two new attributes to the DataFrame that capture for each person their bmi and the respective classification

In [36]:
#your code goes here
df['bmi'] = bmi(df['Weight'], df['HeightM'])
df['bmi_type'] = df['bmi'].apply(bmi_classification)
df.head()

Gender  Height  Weight  HeightM        bmi    bmi_type
0    Male     174      96     1.74  31.708284       Obese
1    Male     189      87     1.89  24.355421      Normal
2  Female     185     110     1.85  32.140248       Obese
3  Female     195     104     1.95  27.350427  Overweight
4    Male     149      61     1.49  27.476240  Overweight

Example output of how your dataframe should look like in the end:

|   | Gender | Height | Weight | HeightM | bmi       | bmi_type   |
|---|--------|--------|--------|---------|-----------|------------|
| 0 | Male   | 174    | 96     | 1.74    | 31.708284 | Obese      |
| 1 | Male   | 189    | 87     | 1.89    | 24.355421 | Normal     |
| 2 | Female | 185    | 110    | 1.85    | 32.140248 | Obese      |
| 3 | Female | 195    | 104    | 1.95    | 27.350427 | Overweight |
| 4 | Male   | 149    | 61     | 1.49    | 27.476240 | Overweight |

## Exercise 2

In this exercise you will download and process a dataset capturing player salaries from a recent NBA season. 

As your first task, download the dataset in `csv` format from the following url:

`https://storage.googleapis.com/comp240-stores/nba_salaries.csv`

Afterwards, answer the 5 questions that follow.

In [37]:
#your code goes here
df = pd.read_csv('https://storage.googleapis.com/comp240-stores/nba_salaries.csv')
df.head()

PLAYER POSITION           TEAM     SALARY
0    Paul Millsap       PF  Atlanta Hawks  18.671659
1      Al Horford        C  Atlanta Hawks  12.000000
2  Tiago Splitter        C  Atlanta Hawks   9.756250
3     Jeff Teague       PG  Atlanta Hawks   8.000000
4     Kyle Korver       SG  Atlanta Hawks   5.746479

### How much money did each team pay for its players’ salaries?

In [48]:
# your code goes here
salaries_per_team = df['SALARY'].groupby(df['TEAM']).sum()
salaries_per_team

TEAM
Atlanta Hawks              69.573103
Boston Celtics             50.285499
Brooklyn Nets              57.306976
Charlotte Hornets          84.102397
Chicago Bulls              78.820890
Cleveland Cavaliers       102.312412
Dallas Mavericks           65.762559
Denver Nuggets             62.429404
Detroit Pistons            42.211760
Golden State Warriors      94.085137
Houston Rockets            85.285837
Indiana Pacers             62.695023
Los Angeles Clippers       66.074113
Los Angeles Lakers         68.607944
Memphis Grizzlies          93.796439
Miami Heat                 81.528667
Milwaukee Bucks            52.258355
Minnesota Timberwolves     65.847421
New Orleans Pelicans       80.514606
New York Knicks            69.404994
Oklahoma City Thunder      96.832165
Orlando Magic              77.623940
Philadelphia 76ers         42.481345
Phoenix Suns               50.520815
Portland Trail Blazers     45.446878
Sacramento Kings           68.384890
San Antonio Spurs          84.652

### Which teams are the top-5 and bottom-5 spenders?

In [62]:
# your code goes here
top_n = 5

print(f'\tBOTTOM {top_n} SPENDERS:\n{salaries_per_team.sort_values(ascending=True)[:top_n]}')
print(f'\tTOP {top_n} SPENDERS:\n{salaries_per_team.sort_values(ascending=False)[:top_n]}')

	BOTTOM 5 SPENDERS:
TEAM
Detroit Pistons           42.211760
Philadelphia 76ers        42.481345
Portland Trail Blazers    45.446878
Boston Celtics            50.285499
Phoenix Suns              50.520815
Name: SALARY, dtype: float64
	TOP 5 SPENDERS:
TEAM
Cleveland Cavaliers      102.312412
Oklahoma City Thunder     96.832165
Golden State Warriors     94.085137
Memphis Grizzlies         93.796439
Washington Wizards        90.047498
Name: SALARY, dtype: float64


### How many NBA players were there in each of the five positions?

In [69]:
# your code goes here
players_by_position = df['POSITION'].groupby(df['POSITION']).count()
players_by_position

POSITION
C     69
PF    85
PG    85
SF    82
SG    96
Name: POSITION, dtype: int64

### What was the average salary of the players at each of the five positions?

In [77]:
# your code goes here
avg_position_salary = df[['POSITION', 'SALARY']].groupby('POSITION').mean()
avg_position_salary

SALARY
POSITION          
C         6.082913
PF        4.951344
PG        5.165487
SF        5.532675
SG        3.988195

### Depict the mean, median, std, min and max SALARY for each of the five positions

In [79]:
# your code goes here
df[['POSITION', 'SALARY']].groupby('POSITION').agg(['mean', 'median', 'std', 'min', 'max'])

SALARY                                         
              mean    median       std       min        max
POSITION                                                   
C         6.082913  4.204200  5.277136  0.111444  22.359364
PF        4.951344  3.300000  4.949479  0.049709  22.192730
PG        5.165487  3.036927  5.038230  0.055722  21.468695
SF        5.532675  2.658953  6.337213  0.030888  25.000000
SG        3.988195  2.233779  4.355577  0.049709  20.000000

## Exercise 3

In this exercise you will download and process a dataset capturing detailed information of the passengers that embarked the Titanic.

As your first task, download the dataset in `csv` format from the following url:

`https://storage.googleapis.com/comp240-stores/titanic.csv`

Afterwards, answer the questions that follow.

In [89]:
#your code goes here
df = pd.read_csv('https://storage.googleapis.com/comp240-stores/titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### How many female and how many male passengers survived?

In [94]:
#your code goes here
df[df['Survived'] == 1][['Survived', 'Sex']].groupby('Sex').count()

Survived
Sex             
female       233
male         109

### Answer the same question but this time return a percentage for each sex against the total number of passengers survived.

In [113]:
#your code goes here
total_passengers_survived = df[df['Survived'] == 1]['Survived'].count()
df_survived = df[df['Survived'] == 1][['Survived', 'Sex']].groupby('Sex').count()
df_survived['Percentage'] = df_survived['Survived'] / total_passengers_survived * 100
df_survived

Survived  Percentage
Sex                         
female       233   68.128655
male         109   31.871345

### What is the percentage of female passengers that survived from the female passengers and what is this percentage for the mail passengers?

In [ ]:
#your code goes here

### Answer the above question by also depicting class-wise (`Pclass` attribute) the survivor rate.

In [ ]:
#your code goes here

### Achieve the same result by embracing a pivot table

In [ ]:
#your code goes here

### Create a new attribute `age_type` that labels each passenger as either (i) `infant` (age 0-3); (ii) `child` (age 4-12); (iv) `teenager` (age 13-17); (iv) `youngster` (age 18-33); (v) `adult` (age 34-59); or (vi) `senior` (age 60 and above)

In [ ]:
#your code goes here

### Redesign the pivot table previously requested to now present the survivor rates for each sex and age type

In [ ]:
#your code goes here